File: 3 course/6 semester/Software development for mobile platforms/films_viewer_kotlin/app/src/main/java/com/example/films_viewer_kotlin/models/Movie.kt
package com.example.films_viewer_kotlin.models

data class Movie(
    val id: String,
    var title: String,
    var description: String,
    var year: Int,
    var duration: Int,
    var genres: List<String>,
    var director: String,
    var imageUrls: List<String>,
    var posterUrl: String,
    var favoritedBy: List<String>
) {
    fun toFirestore(): Map<String, Any> {
        return mapOf(
            "title" to title,
            "description" to description,
            "year" to year,
            "duration" to duration,
            "genres" to genres,
            "director" to director,
            "imageUrls" to imageUrls,
            "posterUrl" to posterUrl,
            "favoritedBy" to favoritedBy
        )
    }

    companion object {
        fun fromFirestore(data: Map<String, Any?>, id: String): Movie {
            return Movie(
                id = id,
                title = data["title"] as? String ?: "",
                description = data["description"] as? String ?: "",
                year = (data["year"] as? Number)?.toInt() ?: 0,
                duration = (data["duration"] as? Number)?.toInt() ?: 0,
                genres = (data["genres"] as? List<*>)?.mapNotNull { it as? String } ?: emptyList(),
                director = data["director"] as? String ?: "",
                imageUrls = (data["imageUrls"] as? List<*>)?.mapNotNull { it as? String } ?: emptyList(),
                posterUrl = data["posterUrl"] as? String ?: "",
                favoritedBy = (data["favoritedBy"] as? List<*>)?.mapNotNull { it as? String } ?: emptyList()
            )
        }
    }
}

File: 3 course/6 semester/Software development for mobile platforms/films_viewer_swift/films_viewer_swift/films_viewer_swift/Models/Movie.swift
import Foundation

struct Movie {
    let id: String
    var title: String
    var description: String
    var year: Int
    var duration: Int
    var genres: [String]
    var director: String
    var imageUrls: [String]
    var posterUrl: String
    var favoritedBy: [String]
    
    init(id: String, data: [String: Any]) {
        self.id = id
        self.title = data["title"] as? String ?? ""
        self.description = data["description"] as? String ?? ""
        self.year = data["year"] as? Int ?? 0
        self.duration = data["duration"] as? Int ?? 0
        self.genres = data["genres"] as? [String] ?? []
        self.director = data["director"] as? String ?? ""
        self.imageUrls = data["imageUrls"] as? [String] ?? []
        self.posterUrl = data["posterUrl"] as? String ?? ""
        self.favoritedBy = data["favoritedBy"] as? [String] ?? []
    }
    
    func toFirestore() -> [String: Any] {
        return [
            "title": title,
            "description": description,
            "year": year,
            "duration": duration,
            "genres": genres,
            "director": director,
            "imageUrls": imageUrls,
            "posterUrl": posterUrl,
            "favoritedBy": favoritedBy
        ]
    }
}
